rfcs/xxxx-v5-content-api.md

```diff
@@ -0,0 +1,217 @@
+| `GET`    | `/api/:contentType`                 | Find document                  |
+| `PUT`    | `/api/:contentType`                 | Create or Update the document  |
+| `DELETE` | `/api/:contentType`                 | Delete document                |
+| `POST`   | `/api/:contentType/actions/:action` | Actions on the single document |
+
+### Add `documentId`
+
+Introduction of the `documentId` field and renaming of the old `id` to `entryId` or `entityId`
+
+**Example**
+
+```tsx
+{
+  "data": {
+    "documentId": "clkgylmcc000008lcdd868feh",
+    "entryId": "clkgylmcc000008lcdd868feh"
+  },
+  "meta": {
+    "pagination": {
+      "page": 1,
+      "pageSize": 10
+    }
+  }
+}
+```
+
+### Introduce a simplified format
+
+We are planning to introduce a new version of the API response that flattens the structure and removes some of the complex nesting from v4.
+
+To avoid conflicts between features & Content Type attributes we will keep the `meta` object but completely flatten the `attributes` & `data.attributes` sub paths.
+
+**Before**
+
+```json
+{
+  "data": {
+    "id": 1,
+    "attributes": {
+      "title": "Article A",
+      "locale": "en",
+      "createdAt": "2023-01-01T00:00:00.000Z",
+      "updatedAt": "2023-01-01T00:00:00.000Z",
+      "publishedAt": "2023-01-01T00:00:00.000Z",
+      "relation": {
+        "data": {
+          "id": 1,
+          "attributes": {
+            "name": "Category A",
+            "locale": "en",
+            "createdAt": "2023-01-01T00:00:00.000Z",
+            "updatedAt": "2023-01-01T00:00:00.000Z",
+            "publishedAt": "2023-01-01T00:00:00.000Z"
+          }
+        }
+      }
+    }
+  },
+  "meta": {
+    "pagination": {
+      "page": 1,
+      "pageSize": 10
+    }
+  }
+}
+```
+
+**After**
+
+```json
+{
+  "data": {
+    "documentId": "clkgylmcc000008lcdd868feh",
+    "entryId": "cpmnyztbcc964008lcft812feo",
+    "title": "Article A",
+    "relation": {
+      "documentId": "clkgylw7d000108lc4rw1bb6s",
+      "entryId": 1,
+      "name": "Category A",
+      "meta": {
+        "locale": "en",
+        "createdAt": "2023-01-01T00:00:00.000Z",
+        "updatedAt": "2023-01-01T00:00:00.000Z",
+        "publishedAt": "2023-01-01T00:00:00.000Z"
+      }
+    },
+    "meta": {
+      "locale": "fr",
+      "createdAt": "2023-01-01T00:00:00.000Z",
+      "updatedAt": "2023-01-01T00:00:00.000Z",
+      "publishedAt": "2023-01-01T00:00:00.000Z"
+    }
+  },
+  "meta": {
+    "pagination": {
+      "page": 1,
+      "pageSize": 10
+    }
+  }
+}
+```
+
+### Separating meta attributes to avoid any current & future name conflicts
+
+```json
+{
+  "data": {
+    "documentId": "clkgylmcc000008lcdd868feh",
+    "title": "Article A",
+    "description": "My description",
+    "meta": {
+      "locale": "",
+      "publishedAt": "",
+      "createdAt": "",
+      "createdBy": ""
+    }
+  }
+}
+```
+
+Pros:
+
+- No conflicts between user & system attributes
```

Comment:
How can there be conflicts anyway? All fields are added flat to the database table, so noone can name a field created_at.

Reply:
The reference is to conflicts from existing projects with a field like "publishedBy" or "release" that do not exist as internal fields today but if upgraded to v5 would be a breaking change to those users and result in a conflict. Likewise if we decide to add any other system/internal fields during the v5 lifecycle.